drivers/gpio/gpio_ambiq.c

```diff
@@ -0,0 +1,353 @@
+{
+	const struct ambiq_gpio_config *const dev_cfg = dev->config;
+
+	for (int i = 0; i < BITS; i++) {
+		if ((pins >> i) & 1) {
+			am_hal_gpio_state_write(i + dev_cfg->pin_offset, AM_HAL_GPIO_OUTPUT_TOGGLE);
+		}
+	}
+
+	return 0;
+}
+
+static void ambiq_gpio_isr(const struct device *dev)
+{
+	struct ambiq_gpio_data *const data = dev->data;
+	const struct ambiq_gpio_config *const dev_cfg = dev->config;
+
+	uint32_t int_status;
+	k_spinlock_key_t key = k_spin_lock(&data->lock);
+
+	am_hal_gpio_interrupt_irq_status_get(dev_cfg->irq_num, false, &int_status);
+	am_hal_gpio_interrupt_irq_clear(dev_cfg->irq_num, int_status);
+
+	gpio_fire_callbacks(&data->cb, dev, int_status);
+	k_spin_unlock(&data->lock, key);
+}
+
+static int ambiq_gpio_pin_interrupt_configure(const struct device *dev, gpio_pin_t pin,
+					      enum gpio_int_mode mode, enum gpio_int_trig trig)
+{
+	const struct ambiq_gpio_config *const dev_cfg = dev->config;
+	struct ambiq_gpio_data *const data = dev->data;
+
+	am_hal_gpio_pincfg_t pincfg;
+	int gpio_pin = pin + dev_cfg->pin_offset;
+	uint32_t int_status;
+	int ret;
+
+	ret = am_hal_gpio_pinconfig_get(gpio_pin, &pincfg);
+
+	if (mode == GPIO_INT_MODE_DISABLED) {
+		pincfg.GP.cfg_b.eIntDir = AM_HAL_GPIO_PIN_INTDIR_NONE;
+		ret = am_hal_gpio_pinconfig(gpio_pin, pincfg);
+		k_spinlock_key_t key = k_spin_lock(&data->lock);
+
+		ret = am_hal_gpio_interrupt_irq_status_get(dev_cfg->irq_num, false, &int_status);
+		ret = am_hal_gpio_interrupt_irq_clear(dev_cfg->irq_num, int_status);
+		ret = am_hal_gpio_interrupt_control(AM_HAL_GPIO_INT_CHANNEL_0,
+						    AM_HAL_GPIO_INT_CTRL_INDV_DISABLE,
+						    (void *)&gpio_pin);
+		k_spin_unlock(&data->lock, key);
+
+	} else {
+		if (mode == GPIO_INT_MODE_LEVEL) {
+			return -ENOTSUP;
+		}
+		switch (trig) {
+		case GPIO_INT_TRIG_LOW:
+			pincfg.GP.cfg_b.eIntDir = AM_HAL_GPIO_PIN_INTDIR_HI2LO;
+			break;
+		case GPIO_INT_TRIG_HIGH:
+			pincfg.GP.cfg_b.eIntDir = AM_HAL_GPIO_PIN_INTDIR_LO2HI;
+			break;
+		case GPIO_INT_TRIG_BOTH:
+			return -ENOTSUP;
+		}
+		ret = am_hal_gpio_pinconfig(gpio_pin, pincfg);
+
+		k_spinlock_key_t key = k_spin_lock(&data->lock);
+
+		ret = am_hal_gpio_interrupt_irq_status_get(dev_cfg->irq_num, false, &int_status);
+		ret = am_hal_gpio_interrupt_irq_clear(dev_cfg->irq_num, int_status);
+		ret = am_hal_gpio_interrupt_control(AM_HAL_GPIO_INT_CHANNEL_0,
+						    AM_HAL_GPIO_INT_CTRL_INDV_ENABLE,
+						    (void *)&gpio_pin);
+		k_spin_unlock(&data->lock, key);
+	}
+	return ret;
+}
+
+static int ambiq_gpio_manage_callback(const struct device *dev, struct gpio_callback *callback,
+				      bool set)
+{
+	struct ambiq_gpio_data *const data = dev->data;
+
+	return gpio_manage_callback(&data->cb, callback, set);
+}
+
+static int ambiq_gpio_init(const struct device *port)
+{
+	const struct ambiq_gpio_config *const dev_cfg = port->config;
+
+	NVIC_ClearPendingIRQ(dev_cfg->irq_num);
+
+	dev_cfg->cfg_func();
+
+	irq_enable(dev_cfg->irq_num);
```

Comment:
is this always required?

Reply:
I moved it to `ambiq_gpio_pin_interrupt_configure`.